src/main/java/uk/gov/hmcts/darts/task/runner/impl/AudioLinkingAutomatedTask.java

```diff
@@ -0,0 +1,107 @@
+import uk.gov.hmcts.darts.task.api.AutomatedTaskName;
+import uk.gov.hmcts.darts.task.config.AutomatedTaskConfigurationProperties;
+import uk.gov.hmcts.darts.task.runner.AutoloadingManualTask;
+import uk.gov.hmcts.darts.task.service.LockService;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@Slf4j
+@Component
+public class AudioLinkingAutomatedTask extends AbstractLockableAutomatedTask
+    implements AutoloadingManualTask {
+
+    private final EventRepository eventRepository;
+    private final MediaRepository mediaRepository;
+    private final MediaLinkedCaseRepository mediaLinkedCaseRepository;
+
+    @Getter
+    private final Integer audioBufferSeconds;
+    private final HearingRepository hearingRepository;
+
+    protected AudioLinkingAutomatedTask(AutomatedTaskRepository automatedTaskRepository,
+                                        AutomatedTaskConfigurationProperties automatedTaskConfigurationProperties,
+                                        LogApi logApi, LockService lockService,
+                                        EventRepository eventRepository,
+                                        MediaRepository mediaRepository, MediaLinkedCaseRepository mediaLinkedCaseRepository,
+                                        @Value("${darts.automated-tasks.audio-linking.audio-buffer-seconds:0}")
+                                        Integer audioBufferSeconds, HearingRepository hearingRepository) {
+        super(automatedTaskRepository, automatedTaskConfigurationProperties, logApi, lockService);
+        this.eventRepository = eventRepository;
+        this.mediaRepository = mediaRepository;
+        this.mediaLinkedCaseRepository = mediaLinkedCaseRepository;
+        this.audioBufferSeconds = audioBufferSeconds;
+        this.hearingRepository = hearingRepository;
+    }
+
+    @Override
+    public AutomatedTaskName getAutomatedTaskName() {
+        return AutomatedTaskName.AUDIO_LINKING_TASK_NAME;
+    }
+
+    @Override
+    @Transactional
+    public void run() {
+        super.run();
+    }
+
+
+    @Override
+    @Transactional
+    protected void runTask() {
+        List<EventEntity> events = eventRepository.findAllByEventStatus(EventStatus.AUDIO_LINK_NOT_DONE_MODERNISED.getStatusNumber(),
+                                                                        Limit.of(getAutomatedTaskBatchSize()));
+        final Set<MediaLinkedCaseEntity> mediaLinkedCaseEntities = new HashSet<>();
+        final Set<HearingEntity> editedHearingEntities = new HashSet<>();
+
+        events.forEach(event -> processEvent(event, editedHearingEntities, mediaLinkedCaseEntities));
+
+        hearingRepository.saveAll(editedHearingEntities);
+        mediaLinkedCaseRepository.saveAll(mediaLinkedCaseEntities);
+        eventRepository.saveAll(events);
+    }
+
+    void processEvent(EventEntity event, Set<HearingEntity> editedHearingEntities, Set<MediaLinkedCaseEntity> mediaLinkedCaseEntities) {
```

Comment:
I don't think we can re-use this logic linkAudioToHearingByEvent is based on the provided courthouse/court room which is not possible here as we need to batch the requests and process all courthouses / courtrooms and this logic exclusively looks at heritage events not modernised.

the mediaLinkedCaseHelper.addCase(savedMedia, hearing.getCourtCase()); methods looks like it could be useful however the approach used in this task is more performant as it does not need to load all MediaLinkedCaseEntity and CourtCaseEntity (as well as any associated eager loaded entities). As well as bulk saving all of which have significant performance benefits. With this in mind there is an argument that we need to raise a tech-debt ticket to improve the performance of  the mediaLinkedCaseHelper.addCase method to avoid the redundant entity loading.

`
 if (!mediaLinkedCaseRepository.existsByMediaAndCourtCase(mediaEntity, courtCase)) {
     mediaLinkedCaseEntities.add(new MediaLinkedCaseEntity(mediaEntity, courtCase));
 }`

Reply:
i think we need to have a chat